src/Compilers/Server/VBCSCompiler/NamedPipeClientConnectionHost.cs

```diff
@@ -13,50 +13,208 @@
+            {
+                Debug.Assert(connection is null || exception is null);
+                NamedPipeClientConnection = connection;
+                Exception = exception;
+            }
+        }
+
+        private CancellationTokenSource? _cancellationTokenSource;
+        private ImmutableArray<Task> _listenTasks;
+        private AsyncQueue<ListenResult>? _queue;
+
+        public string PipeName { get; }
+        public bool IsListening { get; private set; }
 
         internal NamedPipeClientConnectionHost(string pipeName)
         {
-            _pipeName = pipeName;
+            PipeName = pipeName;
+        }
+
+        public void BeginListening()
+        {
+            if (IsListening)
+            {
+                throw new InvalidOperationException();
+            }
+
+            Debug.Assert(_cancellationTokenSource is null);
+            Debug.Assert(_listenTasks.IsDefault);
+            Debug.Assert(_queue is null);
+
+            IsListening = true;
+            _cancellationTokenSource = new CancellationTokenSource();
+            _queue = new AsyncQueue<ListenResult>();
+
+            // The choice of 4 here is a bit arbitrary. The compiler server needs to scale to the number of clients that 
+            // msbuild is going to attempt to connect here and be able to establish each connection in one second. In the 
+            // majority of cases even one is enough to accomplish this. Four though gives us enough wiggle room to handle
+            // severe load scenarios.
+            // 
+            // Should you ever want to change this number in the future make sure to test the new values on sufficiently
+            // large builds such as dotnet/roslyn or dotnet/runtime
+            var listenCount = Math.Min(4, Environment.ProcessorCount);
+            var listenTasks = new List<Task>(capacity: listenCount);
+            int clientLoggingIdentifier = 0;
+            for (int i = 0; i < listenCount; i++)
+            {
+                var task = Task.Run(() => ListenCoreAsync(PipeName, _queue, GetNextClientLoggingIdentifier, _cancellationTokenSource.Token));
+                listenTasks.Add(task);
+            }
+            _listenTasks = ImmutableArray.CreateRange(listenTasks);
+
+            string GetNextClientLoggingIdentifier()
+            {
+                var count = Interlocked.Increment(ref clientLoggingIdentifier);
+                return $"Client{count}";
+            }
+        }
+
+        public void EndListening()
+        {
+            if (!IsListening)
+            {
+                throw new InvalidOperationException();
+            }
+
+            Debug.Assert(_cancellationTokenSource is object);
+            Debug.Assert(_queue is object);
+            Debug.Assert(!_listenTasks.IsDefault);
+
+            _cancellationTokenSource.Cancel();
+            try
+            {
+                Task.WaitAll(_listenTasks.ToArray());
```

Comment:
>_listenTasks.ToArray() [](start = 29, length = 22)

Minor: Consider declaring `_listenTasks` field and `listenTasks` local in `BeginListening` as `Task[]`, to avoid converting in each method.